### ISSUE TITLE
Invalid .h5ad after velocity estimation?

### ISSUE THREAD

**Issue by TobiTekath** [theislab/scvelo#88]
Hi Volker,

apparently I am unable to save and reload my anndata object after velocity estimation. Before velocity estimation (so after _pp.filter_and_normalize_ and _pp.moments_) it still works fine.

## Code:
```
scv.tl.velocity(anndata)

```
> computing velocities
    finished (0:00:14) --> added 
    'velocity', velocity vectors for each individual cell (adata.layers)
```

anndata.write("/home/test.h5ad")
test = scv.read("/home/test.h5ad")
```
## Traceback:
```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-29-e1010dc05ac3> in <module>
----> 1 test = scv.read("/home/test.h5ad")

/miniconda/lib/python3.7/site-packages/scanpy/readwrite.py in read(filename, backed, sheet, ext, delimiter, first_column_names, backup_url, cache, **kwargs)
     95             filename, backed=backed, sheet=sheet, ext=ext,
     96             delimiter=delimiter, first_column_names=first_column_names,
---> 97             backup_url=backup_url, cache=cache, **kwargs,
     98         )
     99     # generate filename and read to dict

/miniconda/lib/python3.7/site-packages/scanpy/readwrite.py in _read(filename, backed, sheet, ext, delimiter, first_column_names, backup_url, cache, suppress_cache_warning, **kwargs)
    497     if ext in {'h5', 'h5ad'}:
    498         if sheet is None:
--> 499             return read_h5ad(filename, backed=backed)
    500         else:
    501             logg.debug(f'reading sheet {sheet} from file {filename}')

/miniconda/lib/python3.7/site-packages/anndata/readwrite/read.py in read_h5ad(filename, backed, chunk_size)
    445     else:
    446         # load everything into memory
--> 447         constructor_args = _read_args_from_h5ad(filename=filename, chunk_size=chunk_size)
    448         X = constructor_args[0]
    449         dtype = None

/miniconda/lib/python3.7/site-packages/anndata/readwrite/read.py in _read_args_from_h5ad(adata, filename, mode, chunk_size)
    500     if not backed:
    501         f.close()
--> 502     return AnnData._args_from_dict(d)
    503 
    504 

/miniconda/lib/python3.7/site-packages/anndata/core/anndata.py in _args_from_dict(ddata)
   2182                         d_true_keys[ann][k_stripped] = pd.Categorical.from_codes(
   2183                             codes=d_true_keys[ann][k_stripped].values,
-> 2184                             categories=v,
   2185                         )
   2186                 k_to_delete.append(k)

/miniconda/lib/python3.7/site-packages/pandas/core/arrays/categorical.py in from_codes(cls, codes, categories, ordered, dtype)
    676         """
    677         dtype = CategoricalDtype._from_values_or_dtype(
--> 678             categories=categories, ordered=ordered, dtype=dtype
    679         )
    680         if dtype.categories is None:

/miniconda/lib/python3.7/site-packages/pandas/core/dtypes/dtypes.py in _from_values_or_dtype(cls, values, categories, ordered, dtype)
    333             # Note: This could potentially have categories=None and
    334             # ordered=None.
--> 335             dtype = CategoricalDtype(categories, ordered)
    336 
    337         return dtype

/miniconda/lib/python3.7/site-packages/pandas/core/dtypes/dtypes.py in __init__(self, categories, ordered)
    224 
    225     def __init__(self, categories=None, ordered: OrderedType = ordered_sentinel):
--> 226         self._finalize(categories, ordered, fastpath=False)
    227 
    228     @classmethod

/miniconda/lib/python3.7/site-packages/pandas/core/dtypes/dtypes.py in _finalize(self, categories, ordered, fastpath)
    345 
    346         if categories is not None:
--> 347             categories = self.validate_categories(categories, fastpath=fastpath)
    348 
    349         self._categories = categories

/miniconda/lib/python3.7/site-packages/pandas/core/dtypes/dtypes.py in validate_categories(categories, fastpath)
    515         if not fastpath and not is_list_like(categories):
    516             msg = "Parameter 'categories' must be list-like, was {!r}"
--> 517             raise TypeError(msg.format(categories))
    518         elif not isinstance(categories, ABCIndexClass):
    519             categories = Index(categories, tupleize_cols=False)

TypeError: Parameter 'categories' must be list-like, was -2147483648
```
## SessionInfo

scvelo==0.1.19.dev47+8d111c8  scanpy==1.4.4  anndata==0.6.22.post1  loompy==2.0.17  numpy==1.17.0  scipy==1.2.1  matplotlib==3.0.2  sklearn==0.21.3  pandas==0.25.1  

## Extra info

According to [this issue](https://github.com/theislab/single-cell-tutorial/issues/7) I tried with a downgraded version of pandas (0.23.4) - but the error still persisted (even when rewriting the object to file).

Any idea how to solve this?

Best,
Tobias

**Comment by VolkerBergen** [comment]
Hmm.. I guess you would have to screen through all `.obs` and `.var`, check their dtype to identify the villain. Also, see whether `scv.utils.strings_to_categorical(adata)` repairs the datatype.

**Comment by TobiTekath** [comment]
Thanks for the quick reply, it helped me to further investigate. 

I found out, that not the `.var` added by  `scv.tl.velocity()` are not the culprits but the internal call to `strings_to_categoricals` which changes the dtypes of some (unrelated) metadata `.obs`.
https://github.com/theislab/scvelo/blob/c2f54cf763f77393ff70af671c29aa9cac2f202b/scvelo/tools/velocity.py#L151

Especially I have some `.obs` which are just 0 for each cell. Before the velocity computation they are stored as `int32` and do not cause any trouble when saving/reading to/from `.h5ad`. By `strings_to_categoricals` they are converted to a categorical with categories of length 1.
Apparently the `.h5ad` reading fails for categoricals with categories of length 1 (at least thats what I observed).

Is would be great if I could keep my `.obs` columns, even if they do not really have any informational value.

Would it be possible to skip the categorical conversion for columns with just one numerical value in it? Or is there really a bug in the `.h5ad` reading/writing for categoricals with only one category?

Here is the dtype output of the `.obs` column which causes the error:

```
Name: test, Length: 10000, dtype: category
Categories (1, int64): [0]
```
